src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -207,33 +284,60 @@ public void pair(APDU apdu) {
+   * hash and verify the card signature.
    *
    * @param apduBuffer the APDU buffer
    * @return the length of the reply
    */
   private short pairStep1(byte[] apduBuffer) {
-    preassignedPairingOffset = -1;
-
-    for (short i = 0; i < (short) pairingKeys.length; i += PAIRING_KEY_LENGTH) {
-      if (pairingKeys[i] == 0) {
-        preassignedPairingOffset = i;
-        break;
-      }
+    // Make sure certificate exisits
+    if (idCertStatus != ID_CERTIFICATE_LOCKED) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
 
-    if (preassignedPairingOffset == -1) {
-      ISOException.throwIt(ISO7816.SW_FILE_FULL);
+    // Override first pairing slot
+    preassignedPairingOffset = 0;
+
+    // Compute ECDH secret
+    crypto.ecdh.init(idKeypair.getPrivate());
+    short pubKeyOff = (short)(ISO7816.OFFSET_CDATA + SC_SECRET_LENGTH);
+    try {
+      crypto.ecdh.generateSecret(apduBuffer, (short) (pubKeyOff + 1), apduBuffer[pubKeyOff], secret, (short) 0);
+    } catch(Exception e) {
+      ISOException.throwIt(ISO7816.SW_WRONG_DATA);
+      return 0;
     }
 
-    crypto.sha256.update(pairingSecret, (short) 0, SC_SECRET_LENGTH);
-    crypto.sha256.doFinal(apduBuffer, ISO7816.OFFSET_CDATA, SC_SECRET_LENGTH, apduBuffer, (short) 0);
-    crypto.random.generateData(secret, (short) 0, SC_SECRET_LENGTH);
-    Util.arrayCopyNonAtomic(secret, (short) 0, apduBuffer, SC_SECRET_LENGTH, SC_SECRET_LENGTH);
+    // Hash client salt and ECDH secret and store in secret buffer
+    // secretHash = sha256(clientSalt, ECDH secret)
+    crypto.sha256.update(apduBuffer, ISO7816.OFFSET_CDATA, SC_SECRET_LENGTH);
+    crypto.sha256.doFinal(secret, (short) 0, SC_SECRET_LENGTH, secret, (short) 0);
+
+    // Generate random card salt, and copy to response buffer
+    crypto.random.generateData(apduBuffer, (short) 0, SC_SECRET_LENGTH);
+
+    // Copy card certificate to response buffer
+    Util.arrayCopyNonAtomic(idCertificate, (short) 0, apduBuffer, SC_SECRET_LENGTH, (short) CERTIFICATE_LEN);
+
+    // Sign the secret hash, and copy the signature into the response buffer
+    eccSig.init(idKeypair.getPrivate(), Signature.MODE_SIGN);
+    short len = eccSig.signPreComputedHash(secret, (short) 0, (short) (SC_SECRET_LENGTH), apduBuffer, (short) (SC_SECRET_LENGTH + CERTIFICATE_LEN));
 
-    return (SC_SECRET_LENGTH * 2);
+    // Zero-pad end of signature field (ECDSA sig is variable length, but we send full max size field)
+    short padCount = (short) (ECDSA_MAX_LEN - len);
+    Util.arrayFillNonAtomic(apduBuffer, (short) (SC_SECRET_LENGTH + CERTIFICATE_LEN + len), padCount, (byte) 0);
+
+    // Compute the expected client cryptogram, by hashing the card salt and secret hash. Save in secret buffer.
+    // expectedCryptogram = sha256(cardSalt, secretHash)
+    crypto.sha256.update(apduBuffer, (short) 0, SC_SECRET_LENGTH);
```

Comment:
Didn't you already do this on L317? I'm losing track of what's in your buffer lol

Reply:
That was to generate the "challenge message" in pairStep1. This is to generate the "cryptogram" in pairStep2. The cryptogram is similar to an HMAC on the challenge message, but they are distinct

Reply:
Just wanted you to double check ;)